R/stroke.R

```diff
@@ -0,0 +1,269 @@
+
+  # split the edges into their constituent points
+  edge_pts <- sfheaders::sfc_to_df(edges_sfc)
+
+  # find unique points ("nodes") and assign them IDs
+  nodes <- unique_nodes(edge_pts)
+
+  # build array of line segments, referring to points using their IDs
+  segments <- to_line_segments(edge_pts, nodes)
+
+  # build connectivity table: for each node, find intersecting line segments
+  links <- get_links(segments)
+
+  # calculate interior angles between segment pairs, identify best links
+  best_links <- best_link(nodes, segments, links, angle_threshold)
+
+  # verify that the best links identified fulfill input requirements
+  final_links <- cross_check_links(best_links, flow_mode)
+
+  # merge line segments into strokes following the predetermined connectivity
+  strokes  <- merge_lines(nodes, segments, final_links, from_edge)
+
+  # add the CRS to the edges, done!
+  sf::st_crs(strokes) <- crs
+  return(strokes)
+}
+
+#' Find unique nodes and label them with IDs
+#' @noRd
+unique_nodes <- function(edge_points) {
+  nodes <- dplyr::distinct(edge_points, x, y)
+  nodes$node_id <- seq_len(nrow(nodes))
+  return(nodes)
+}
+
+#' @noRd
+to_line_segments <- function(points, nodes) {
+  # label all the points using the node IDs
+  points <- dplyr::left_join(points, nodes, by = c("x", "y"))
+
+  # use the attribute `linestring_id` to identify edge start- and end-points
+  is_startpoint <- !duplicated(points$linestring_id)
+  is_endpoint <- !duplicated(points$linestring_id, fromLast = TRUE)
+  # we build the array of line segments, with shape (nsegements, 2)
+  # values are the node IDs
+  start <- points[!is_endpoint, "node_id"]
+  end <- points[!is_startpoint, "node_id"]
+  segments <- cbind(start, end)
+  return(segments)
+}
+
+#' @noRd
+get_links <- function(segments) {
+  nsegments <- nrow(segments)
+  connections <- data.frame(node_id = as.vector(segments)) |>
+    dplyr::group_by(node_id) |>
+    dplyr::group_rows()  |>
+    lapply(function(x) (x - 1) %% nsegments + 1)
+  return(connections)
+}
+
+#' @noRd
+best_link <- function(nodes, segments, links, angle_threshold = 0) {
+
+  get_linked_segments <- function(segment_id, node_id) {
+    # find the segments connected to the given one via the given node
+    # 1. find all segments connected to the node
+    segs <- links[[node_id]]
+    # 2. exclude the given segment from the list
+    is_current_segment <- segs == segment_id
+    linked_segments <- segs[!is_current_segment]
+    return(linked_segments)
+  }
+
+  get_linked_nodes <- function(node_id, segment_id) {
+    # find the node connected to the given one via the given segment(s)
+    # 1. get the nodes that are part of the given segment(s)
+    nds <- segments[segment_id, ]
+    # 2. flatten the array row by row (i.e. along the node dimension)
+    nds <- as.vector(t(nds))
+    # 3. exclude the given node from the list
+    is_current_node <- nds %in% node_id
+    linked_nodes <- nds[!is_current_node]
+    return(linked_nodes)
+  }
```

Comment:
These nested functions should be move outside the main function. The data they use, e.g., `link` in the case of `get_linked_segments`, can become an additional argument so that the functions are self-contained and do not depend on objects in their enclosing environment. I also see some repetition with functions in `merge_lines`.

Reply:
I have pull the functions out, adding the data that is sliced within them as an additional argument. I am not 100% sure whether R passes by value or by reference, so whether this is actually increasing memory usage. However, slicing the arrays on input would hinder readability of some of these functions - will open an issue to address this and the duplicated content with `merge_lines`